string/string.h

```diff
@@ -0,0 +1,267 @@
+        }
+
+        if (i < end) {
+            for (; i < end; i++) {
+                this->string[i] = *character++;
+                this->len++;
+            }
+
+            this->string[this->len] = NULL_TERMINATOR;
+        }
+    }
+
+    /**
+     * Insert a character at a certain index
+     * @param position
+     * @param character
+     */
+    void insert(int position, char character) {
+        if (position > len || len + 1 >= size) return;
+
+        for (int i = len + 2; i > position - 1; i--) {
+            this->string[i] = this->string[i - 1];
+        }
+
+        this->string[position] = character;
+        this->len++;
+    }
+
+    /**
+     * Removes character at position specified
+     * @param position to remove at
+     */
+    void remove(int position) {
+        if (position > len) return;
+
+        for (int i = position; i < len + 1; i++) {
+            this->string[i] = this->string[i + 1];
+        }
+        this->len--;
+    }
+
+    /**
+     * Wipes the entire string buffer
+     */
+    void clear() {
+        int i = 0;
+        while (this->len > 0) {
+            string[i++] = NULL_TERMINATOR;
+            this->len--;
+        }
+    }
+
+    /**
+     * Recount the length of the string in case
+     * null gets randomly added to a string
+     *
+     * @return New Length or -1 if unchanged
+     */
+    int recount() {
+        int len_count = 0;
+
+        while (this->string[(len_count++) + 1] != NULL_TERMINATOR);
+
+        if (len_count != this->len) {
+            this->len = len_count;
+
+            return this->len;
+        }
+
+        return -1;
+    }
+
+    /**************************************
+     * Iterators
+     **************************************/
+     // TODO: Could be nice. Maybe add later after getting base functionality down.
+
+    /**************************************
+     * Capacity
+     **************************************/
+
+    /**
+     * Check if string is empty
+     * @return
+     */
+    bool empty() const {
+        return this->len == 0;
+    }
+
+    /**
+     * Check if string fills entire buffer
+     * @return
+     */
+    bool full() const {
+        return this->len >= size;
+    }
+
+    /**
+     * Get the max size of the buffer
+     * @return
+     */
+    size_t max_size() const {
+        return size;
+    }
+
+    /**
+     * Get the remaining available space in the buffer
+     * @return
+     */
+    size_t available() const {
+        return size - len - 1;
+    }
+
+    /**************************************
+     * C API Interfacing
+     **************************************/
+
+    /**
+     * Checks if there is extra space in the buffer
+     * @return boolean on if there's extra space
+     */
+    bool is_trunc() {
+        return this->len < size;
+    }
+
+    /**
+     * Converts remaining characters in buffer to \0
+     */
+    void init_free_space() {
+        // TODO: Use memset or bad for embedded?
```

Comment:
memset is probably okay but for now we're going to avoid using libc to see if we can get away with it